packages/autorest.python/autorest/m4reformatter/__init__.py

```diff
@@ -403,6 +404,98 @@ def to_lower_camel_case(name: str) -> str:
     return re.sub(r"_([a-z])", lambda x: x.group(1).upper(), name)
 
 
+def update_overload_section(
+    overload: Dict[str, Any],
+    yaml_data: Dict[str, Any],
+    section: str,
+):
+    for overload_s, original_s in zip(overload[section], yaml_data[section]):
+        if overload_s.get("type"):
+            overload_s["type"] = original_s["type"]
+        if overload_s.get("headers"):
+            for overload_h, original_h in zip(
+                overload_s["headers"], original_s["headers"]
+            ):
+                if overload_h.get("type"):
+                    overload_h["type"] = original_h["type"]
+
+
+def add_overload(
+    yaml_data: Dict[str, Any], body_type: Dict[str, Any], for_flatten_params=False
+):
+    overload = copy.deepcopy(yaml_data)
+    overload["isOverload"] = True
+    overload["bodyParameter"]["type"] = body_type
+    overload["bodyParameter"]["defaultToUnsetSentinel"] = False
+    overload["overloads"] = []
+    if yaml_data.get("initialOperation"):
+        overload["initialOperation"] = yaml_data["initialOperation"]
+
+    if for_flatten_params:
+        overload["bodyParameter"]["flattened"] = True
+    else:
+        overload["parameters"] = [
+            p for p in overload["parameters"] if not p.get("inFlattenedBody")
+        ]
+    # for yaml sync, we need to make sure all of the responses, parameters, and exceptions' types have the same yaml id
+    for overload_p, original_p in zip(overload["parameters"], yaml_data["parameters"]):
+        overload_p["type"] = original_p["type"]
+    update_overload_section(overload, yaml_data, "responses")
+    update_overload_section(overload, yaml_data, "exceptions")
+
+    # update content type to be an overloads content type
+    content_type_param = next(
+        p for p in overload["parameters"] if p["restApiName"].lower() == "content-type"
+    )
+    content_type_param["inOverload"] = True
+    content_type_param["inDocstring"] = True
+    body_type_description = get_body_type_for_description(overload["bodyParameter"])
+    content_type_param[
+        "description"
+    ] = f"Body Parameter content-type. Content type parameter for {body_type_description} body."
+    content_types = yaml_data["bodyParameter"]["contentTypes"]
+    if body_type["type"] == "binary" and len(content_types) > 1:
+        content_types = "'" + "', '".join(content_types) + "'"
+        content_type_param["description"] += f" Known values are: {content_types}."
+    overload["bodyParameter"]["inOverload"] = True
+    for parameter in overload["parameters"]:
+        parameter["inOverload"] = True
+        parameter["defaultToUnsetSentinel"] = False
+    return overload
+
+
+def add_overloads_for_body_param(yaml_data: Dict[str, Any]) -> None:
```

Comment:
i think i already left this comment by why are we duplicating code that we used to share in preprocesser? the same code is now in emitter and m4reformatter and i'm not sure why